src/uct/ib/base/ib_iface.c

```diff
@@ -930,6 +1007,8 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
 
     self->addr_size  = uct_ib_address_size(self->addr_type);
 
+    kh_init_inplace(uct_ib_ah, &self->ah_hash);
```

Comment:
I guess the cache can grow to number of nodes ? Does it make sense to limit the size of the cache for memory scalability reasons ?

Reply:
Maybe remove the AH from the cache if refcount == PPN?

Reply:
@evgeny-leksikov You can think about some sort of LRU policy once you hit the threshold.

Reply:
PPN is not enough: flow control requires AH too (originally this PR was defined to fix FC slowness).
Some transports creates own iface per EP, we can't stop caching after PPN refcount

what is problem in huge number of nodes? 100k nodes is really big cluster - it will produce 100k AH's, without this PR every DC EP will produce own AH too - nothing changed

Reply:
@hoopoepg ok PPN*2 for TLs which support FC. khash is not for free.
@shamisp knowledge when hash entry is not needed may help in stress cases do not reach the threshold, so i didn't mean that this is a replacement for LRU policy, more likely addition.

anyway, the patch solves the real issue, I think hypothetical limitation at scale should not be a stopper.

Reply:
is there any way to know PPN?